ui/site/src/credentiallessIframes.ts

```diff
@@ -0,0 +1,13 @@
+// Ensure all iframes created by 3rd-party scripts like Twitter or hcaptcha get the `credentialless` attribute.
+export const patchCreateDocumentToEnsureAllIFramesAreCredentialless = () => {
+  if (!(window.crossOriginIsolated && 'credentialless' in window)) return;
```

Comment:
`window.crossOriginIsolated && 'credentialless' in window` is true on every page in my chrome browser. Which means the following code runs for me on every page, and replaces the `document.createElement` function.

I expect that function to be called many, many times, as snabbdom creates and updates the UI. And I'm a bit concerned about the performance hit during page load and gameplay.

Could it be possible to run it on actual iframes element only?

Reply:
I think a mutation observer could work. Watch the parent where we know the iframe is going to be added, and then add the attribute at that point before the src loads.

But it might be a good idea to just disable credentialless in the next deploy so we have more time to test whether the monkey patch or mutation observer truly fixes the hcaptcha issue.

Reply:
ok the setting is back I'll flip it after deploy.

Reply:
Running it on every page was kinda the point, to ensure it catches all iframes and avoid the need for special exceptions. But the function indeed would run fairly often. I would be surprised though if the performance impact would be noticeable. It hardly does anything after all and it's not like we're not running a bunch of other JS around all the elements we create. But I guess one can argue whether the benefit justifies the risk.

I don't really see how one could selectively override the function only for iframes or do it any other way to only target iframes. I doubt a general mutation observer would be more performant and if we'd limit it to only the places where an hcaptcha will appear, I wonder if it wouldn't just be simpler to disable the credentialless coep on those pages.

Although I guess a mutation observer on `.h-captcha` (for example) would ensure we don't miss any, especially if more get added in the future.

It still wouldn't catch other stray iframes though.

Reply:
I would prefer a mutation observer, yes. Mutating the native `document.createElement` function feels way too hackish.

Or can we manually call a function where and when we create an iframe? We'll forget it sometimes, then we'll add it :shrug:

Reply:
We could also find out if adding credentialless attribute to the iframe will work before settling on an injection method.

I must have a very trustworthy IP because I have not yet been able to trigger the hcaptcha through external means.

Reply:
The only thing I can think of why Trevlar is able to repro it is if the embedding hostname was factored into the fingerprint mechanism somehow. He's often using those ephemeral gitpod instances.

Reply:
Do we have any hcaptcha creds laying around for lichess.dev? Or will somebody need to actually register it to reproduce and verify this patch outside of lichess.org

Reply:
Just verified that manually injecting the credentialless attribute before the iframe src is loaded will fix the captcha issue on Chrome, so good news.

Reply:
https://github.com/lichess-org/lila/pull/15171/

We could take this targeted approach for hcaptcha, but I think Benedikt's monkey patch would not hurt performance at all if we wanted something for all iframes. The additional delegation overhead wouldn't be measurable unless it happened hundreds of thousands of times, and that would be quite a DOM

Reply:
> Or can we manually call a function where and when we create an iframe?

The issue is that we aren't creating these iframes ourselves, they are created by third-party scripts from hcaptcha or Twitter.

@schlawg's more targeted PR also seems reasonable though. It does mean we'll need to add something similar anytime we add more such third-party scripts that create iframes but that hopefully should be quite rare. And there's hope that we'll get a better way at some point, like a header that makes the browser make all iframes credentialless automatically.

Reply:
There's no reason the hcaptcha people couldn't put `if ('credentialless' blah blah) iframe.setAttribute('credentialless','')` directly in their script, it's literally one line.

I was thinking about it some more and patching createElement is kinda scary. It's OK if we're the only ones who do it and this is the only place it's done, but as soon as there's more than one patch, indeterminate scheduling means you can't be sure which version of createElement you'll be using, or which one you'll end up when the last callback restores what they consider to be the original (which may itself have been a patch).

Reply:
Excellent point against patching `createElement`.

Reply:
Alas, mutation observer access is too late - the iframe's security context is already frozen at that point. hcaptcha javascript API allows one to take control of calling render with a render=explicit query param. I tried using that to render to a detached div and manipulate the iframe then before adding the div back to the DOM. When I did that, hcaptcha seems to load OK but then freezes for some reason later on. I don't know why.

If we want to solve this with a 'credentialless' attribute in the iframe, patching createElement is the only method I have seen work. In #15171, document.createElement is first reassigned after the top level statements in the hcaptcha.com/1/api.js script have executed. It is then reverted to the stored version when hcaptcha calls its onload callback (this one is provided as a query param). So I've done all I can to minimize the duration of time that document.createElement is dirty. If this sounds nuts (and it kinda does), let's just disable credentialless altogether on clas, magic link, register pages.

Reply:
> indeterminate scheduling means you can't be sure which version of createElement you'll be using

I don't see why this matters. The patch is ordering independent.

> or which one you'll end up when the last callback restores what they consider to be the original (which may itself have been a patch).

That's a legitimate issue, but only if you're restoring the function, which I wasn't doing.

> mutation observer access is too late - the iframe's security context is already frozen at that point.

If you get the iframe to reload, the credentialless should apply. But idk if hcaptcha would be happy with that.

Another option would be to load hcaptcha in an iframe which already has credentialless set. You can directly set the iframe content with `srcdoc`. And any iframe created in there will forcefully inherit the credentiallessness. But again not sure if hcaptcha will be happy with that and I guess this probably also would need extra code to extract the hcaptcha result.

I think your PR is fine and less brittle than those alternatives. Though just disabling credentialless on those pages indeed might be the cleanest solution. Maybe this would even lead to less users seeing the captchas because hcaptcha can then read earlier stored context and decide the user is already trusted. But not sure if hcaptcha actually does stuff like that and seeing a captcha is already extremely rare anyways.

Reply:
> I don't see why this matters. The patch is ordering independent.

It only matters when monkey patching from different modules. Imagine we apply our patch right after html parse. A short time later node module xxyy overwrites the original createElement (which was actually ours) in favor of their patch which they swear is compatible with the original createElement, but uses innerHTML or some other hack to actually create the element (which doesn't contain our goodies). After that, hcaptcha script runs but uses xxyy's patch not ours. I almost feel like this failure would be worth it as we'd know that xxyy needs to be fired into the sun, but there it is.